### PR TITLE
ID-1029 Fulfilment point category

### PR DESCRIPTION
**Pull request by mickeylee25**
This PR includes:
1. Update fulfilment point category fragments and queries.
2. Download the latest schema for queries of fulfilment point category.
3. Update `getMyOrders` query in order to resolve the compiling error.
4. Implement the fulfilment point category component.

We decided to create our own model with protocol-based object originally. However, conforming necessary ApolloType to exposed protocols is way too complicated to handle as there are some limitations around type such as below instance. 
```
protocol Category {
    var id: String { get }
    var translations: [Translation]? { get }
}

extension ApolloType.FragmentFulfilmentPointCategory: Category { } // --> Error: Type 'ApolloType.FragmentFulfilmentPointCategory' does not conform to protocol 'Category'

protocol Translation {
    var language: Language? { get }
    var title: String? { get }
}

extension ApolloType.FragmentFulfilmentPointCategory.Translation: Translation { } //  --> Error: Type 'ApolloType.FragmentFulfilmentPointCategory.Translation' does not conform to protocol 'Translation'

protocol Language {
    var rawValue: String { get }
}

extension ApolloType.Language: Language { }
```
There's an error when conforming the ApolloType to SDK's protocol and we will need to rename the `translations: [Translation]?` after resolving the error. In conclusion, I decide to switch the solution to create SDK's own wrapper like the model `FulfilmentPointCategory`. Although there are more duplicated codes in the callback of GraphQLManager, this would be the easier approach in this phase and this is more consistent with the Android's solution.

### CODE REVIEW

File: Core/GraphQL/GraphQLError.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//
+//  SellError.swift
```
Comment (thisolivier) [inline Core/GraphQL/GraphQLError.swift:2]: Please update to GraphQLError

- [ ] ranem SellError -> GraphQLError

File: Core/GraphQL/GraphQLFiles/sell_fragment_order.graphql [hunk 2]
```diff
@@ -10,7 +10,7 @@ fragment FragmentOrder on Order {
     translations {
       language
       title
-      desciption
+      description
```
Comment (SoaringEarth) [inline Core/GraphQL/GraphQLFiles/sell_fragment_order.graphql:13]: Nice that BE learned to spell

File: Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        page: Int,
+        callback: @escaping (Result<PaginatedObject<FulfilmentPointCategory>, Error>) -> Void
+    ) {
+        graphQLManager.dispatch(
+            query: ApolloType.GetFulfilmentPointCategoriesQuery(pageSize: pageSize, page: page),
+            cachePolicy: .returnCacheDataAndFetch
+        ) { result in
+            switch result {
+            case .success(let response):
+                let items = response.data?.getFulfilmentPointCategories?.edges?.compactMap {
```
Comment (thisolivier) [inline Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift:33]: I'm thinking about the upcoming error handling we're going to have to deal with & that under the success case, we'll have to check for our server's errors, and possibly we should be doing that/thinking about a pattern now.

- [ ] let me know what you think

File: Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    ) {
+        graphQLManager.dispatch(
+            query: ApolloType.GetFulfilmentPointCategoryByIdQuery(id: id),
+            cachePolicy: .returnCacheDataAndFetch
+        ) { result in
+            switch result {
+            case .success(let response):
+                guard
+                    let fulfilmentPointCategory = FulfilmentPointCategory(
+                        response: response.data?.getFulfilmentPointCategory?.fragments.fragmentFulfilmentPointCategory)
```
Comment (thisolivier) [inline Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift:58]: Same concern as above when it comes to us not checking for errors in the success case.

**Review by thisolivier — CHANGES_REQUESTED** [review]
Some small changes & one point to consider on `.success` case errors.

File: Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift (hunk 3 above)
Comment (mickeylee25) [reply Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift:33]: I have the same concern about having error handling in the success case. I can check if the `data` is existed in the `GraphQLResult` in GraphQLManager.
```
private func processGraphQLResult<Data: GraphQLSelectionSet>(
        _ result: Result<GraphQLResult<Data>, Error>,
        completion: @escaping (Result<Data, Error>) -> Void
    ) {
        switch result {
        case .success(let response):
            guard
                let data = response.data,
                !data.jsonObject.isEmpty
            else {
                return completion(.failure(GraphQLError.noDataError))
            }
            completion(.success(data))
        case .failure(let error):
            completion(.failure(error))
        }
    }
```

So the callback of the GraphQLManageable would be:
```
public func dispatch<Query: GraphQLQuery>(
        query: Query,
        cachePolicy: GraphNetworkCachePolicy,
        completion: @escaping (Result<Query.Data, Error>) -> Void
    )
```

However, we still have to safe unwrap the returned Data as the query/mutation type like `getFulfilmentPointCategory` is optional in Apollo SDK and I couldn't find a generic way to safe unwrap the fragments and the normal type. That's the reason why GraphQL is not friendly to FE from my opinion.
```
guard
    let fulfilmentPointCategory = FulfilmentPointCategory(
        response: response.getFulfilmentPointCategory?.fragments.fragmentFulfilmentPointCategory)
else {
    return callback(.failure(GraphQLError.noDataError))
}
```

Anyway, do you think we need to implement this? We convert the result type to `Result<Query.Data, Error>` in GraphQLManager, so that it will throw the empty error when there's no data to map first. And in each repository layer, we still have to safe unwrap the response and throw the empty error under the success case. Really redundant solution honestly. Welcome any more generic solution

Comment (thisolivier) [reply Sell/FulfilmentPoint/Repository/FulfilmentPointRepository.swift:33]: I thought about using the Manager to handle some of this, but I don't think it matches the way GraphQL behaves. Reason being, we can get data AND have an error, or multiple errors, in the success case. It looks something like:

--> Get response, wrap in a result
       --> [failure] We have a single non-GraphQL error --> do normal error handling
       --> [success] We need to check if we have errors and if we have data
             --> [Error and no data] Do error handling as normal?
             --> [Multiple errors and no data] ???
             --> [Errors AND some data] ???
             
Basically, having errors and having data aren't exclusive states in GraphQL, so we can't just catch errors in the successful response and map them back into a Result type. Maybe we could do something like this, but I think there must be something more elegant:

```swift
enum GraphQLResponse {
    case totalSuccess(Query.Data)
    case partialSuccess(Query.Data, [GraphQLError])
    case totalFailure([GraphQLError])
    case serverNeverReached(Error) // This is the current result.failure state
}
```

**Review by thisolivier — APPROVED** [review]

**Review by You-Hsuan — APPROVED** [review]
